src/Futhark/CodeGen/ImpGen/GPU/SegRed.hs

```diff
@@ -73,6 +83,28 @@ maxNumOps = 10
 -- indexes into that t'SubExp' for reading the result.
 type DoSegBody = ([(SubExp, [Imp.TExp Int64])] -> InKernelGen ()) -> InKernelGen ()
 
+-- | Datatype used to distinguish between the different types of reductions we
+-- generate code for. Actually, we only distinguish between non-commutative
+-- reductions with all-Prim parameters, and all other reductions. The former
+-- requires a different (and larger) set of intermediate arrays (in local mem)
+-- and a more elaborate stage one loop body.
+--
+-- TODO (1): ideally this choice would be made implicitly throughout the module,
+-- somehow. Perhaps Troels has a good idea on how to do this elegantly (?).
```

Comment:
I also considered that. My reason for deciding not to do that was that it would mean computing it independently in 3-4 different places in the module, and if in the future we need to distinguish between more than 2 different kinds of reductions, then this becomes spaghetti. I'll think about it!

Reply:
I decided to remove `SegRedKind` for the reason that if it should become beneficial, then it is easy to reinstate, but as is it serves no purpose